# System features (CPU and GPU)

In [2]:
!sudo apt-get install neofetch
!neofetch

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  neofetch
0 upgraded, 1 newly installed, 0 to remove and 11 not upgraded.
Need to get 74.8 kB of archives.
After this operation, 365 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 neofetch all 3.4.0-1 [74.8 kB]
Fetched 74.8 kB in 1s (137 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package neofetch.
(Reading database ... 144628 files and directories current

In [3]:
!nvidia-smi

Mon Nov  9 15:38:02 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 455.32.00    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   48C    P8     8W /  75W |      0MiB /  7611MiB |      0%      Default |
|                               |                      |                 ERR! |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [4]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU @ 2.30GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2300.000
cache size	: 46080 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips	: 4600.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:

processor	:

## Test 1

In [5]:
import time, torch

bas = time.time()

a = torch.rand(10000, 10000, device=torch.device("cuda"))
b = torch.rand(10000, 10000, device=torch.device("cuda"))
c = torch.rand(10000, 10000, device=torch.device("cuda"))
d = torch.rand(10000, 10000, device=torch.device("cuda"))


for i in range(0,1000):
  y = a@b
  z = c@d

  x = y@z

son = time.time()

In [6]:
x

tensor([[6.2079e+10, 6.1564e+10, 6.1465e+10,  ..., 6.1224e+10, 6.2216e+10,
         6.2389e+10],
        [6.2956e+10, 6.2433e+10, 6.2333e+10,  ..., 6.2088e+10, 6.3094e+10,
         6.3269e+10],
        [6.2527e+10, 6.2008e+10, 6.1909e+10,  ..., 6.1666e+10, 6.2665e+10,
         6.2839e+10],
        ...,
        [6.3534e+10, 6.3007e+10, 6.2906e+10,  ..., 6.2659e+10, 6.3674e+10,
         6.3851e+10],
        [6.3003e+10, 6.2480e+10, 6.2380e+10,  ..., 6.2135e+10, 6.3142e+10,
         6.3317e+10],
        [6.2310e+10, 6.1792e+10, 6.1693e+10,  ..., 6.1451e+10, 6.2447e+10,
         6.2620e+10]], device='cuda:0')

In [7]:
"1.test result (second) : " + str(son-bas)

'1.test result (second) : 1071.427838563919'

In [8]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243


In [9]:
!pip install git+git://github.com/andreinechaev/nvcc4jupyter.git
%load_ext nvcc_plugin

  Cloning git://github.com/andreinechaev/nvcc4jupyter.git to /tmp/pip-req-build-g25m6mg7
  Running command git clone -q git://github.com/andreinechaev/nvcc4jupyter.git /tmp/pip-req-build-g25m6mg7
  Created wheel for NVCCPlugin: filename=NVCCPlugin-0.0.2-cp36-none-any.whl size=4307 sha256=3d29f17033659522793af3e9f128bcb0241b338e112bdc554f453870e7a24758
  Stored in directory: /tmp/pip-ephem-wheel-cache-8pk7o_or/wheels/10/c2/05/ca241da37bff77d60d31a9174f988109c61ba989e4d4650516
Successfully built NVCCPlugin
created output directory at /content/src
Out bin /content/result.out


In [10]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu

--2020-11-09 16:06:25--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.217.104.52
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.217.104.52|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1155 (1.1K) [application/cu-seeme]
Saving to: ‘kernel.cu’

kernel.cu           100%[===================>]   1.13K  --.-KB/s    in 0s      

2020-11-09 16:06:26 (75.4 MB/s) - ‘kernel.cu’ saved [1155/1155]



In [12]:
!cat kernel.cu

#include <math.h>
#include <iostream>
#include "cuda_runtime.h"
#include "kernel.h"
#include <stdlib.h>

using namespace std;

__global__ void matrixMultiplicationKernel(float* A, float* B, float* C, int N) {

    int ROW = blockIdx.y*blockDim.y+threadIdx.y;
    int COL = blockIdx.x*blockDim.x+threadIdx.x;

    float tmpSum = 0;

    if (ROW < N && COL < N) {
        // each thread computes one element of the block sub-matrix
        for (int i = 0; i < N; i++) {
            tmpSum += A[ROW * N + i] * B[i * N + COL];
        }
    }
    C[ROW * N + COL] = tmpSum;
}


void matrixMultiplication(float *A, float *B, float *C, int N){

    // declare the number of blocks per grid and the number of threads per block
    // use 1 to 512 threads per block
    dim3 threadsPerBlock(N, N);
    dim3 blocksPerGrid(1, 1);
        if (N*N > 512){
            threadsPerBlock.x = 512;
            threadsPerBlock.y = 512;
            blocksPerGrid.x = ceil(double(N)/double(threadsPerBlock.x));
         

In [13]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h

--2020-11-09 16:09:23--  https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.162.51
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.162.51|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1997 (2.0K) [text/x-chdr]
Saving to: ‘dev_array.h’

dev_array.h         100%[===================>]   1.95K  --.-KB/s    in 0s      

2020-11-09 16:09:24 (121 MB/s) - ‘dev_array.h’ saved [1997/1997]



In [14]:
!cat dev_array.h

#ifndef _DEV_ARRAY_H_
#define _DEV_ARRAY_H_

#include <stdexcept>
#include <algorithm>
#include <cuda_runtime.h>

template <class T>
class dev_array
{
// public functions
public:
    explicit dev_array()
        : start_(0),
          end_(0)
    {}

    // constructor
    explicit dev_array(size_t size)
    {
        allocate(size);
    }
    // destructor
    ~dev_array()
    {
        free();
    }

    // resize the vector
    void resize(size_t size)
    {
        free();
        allocate(size);
    }

    // get the size of the array
    size_t getSize() const
    {
        return end_ - start_;
    }

    // get data
    const T* getData() const
    {
        return start_;
    }

    T* getData()
    {
        return start_;
    }

    // set
    void set(const T* src, size_t size)
    {
        size_t min = std::min(size, getSize());
        cudaError_t result = cudaMemcpy(start_, src, min * sizeof(T), cudaMemcpyHostToDevice);
        if (result != cudaSuccess)
        {
     

In [15]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.h

--2020-11-09 16:10:55--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.177.203
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.177.203|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 112 [text/x-chdr]
Saving to: ‘kernel.h’

kernel.h            100%[===================>]     112  --.-KB/s    in 0s      

2020-11-09 16:10:55 (3.30 MB/s) - ‘kernel.h’ saved [112/112]



In [16]:
!cat kernel.h

#ifndef KERNEL_CUH_
#define KERNEL_CUH_

void matrixMultiplication(float *A, float *B, float *C, int N);

#endif

In [17]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu

--2020-11-09 16:11:13--  https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.217.72.28
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.217.72.28|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1060 (1.0K) [application/cu-seeme]
Saving to: ‘matrixmul.cu’

matrixmul.cu        100%[===================>]   1.04K  --.-KB/s    in 0s      

2020-11-09 16:11:14 (67.0 MB/s) - ‘matrixmul.cu’ saved [1060/1060]



In [18]:
!cat matrixmul.cu

#include <iostream>
#include <vector>
#include <stdlib.h>
#include <time.h>
#include <cuda_runtime.h>
#include "kernel.h"
#include "kernel.cu"
#include "dev_array.h"
#include <math.h>
#include <stdio.h>

using namespace std;

int main()
{
    // Perform matrix multiplication C = A*B
    // where A, B and C are NxN matrices
    int N = 10000;
    int SIZE = N*N;

    // Allocate memory on the host
    vector<float> h_A(SIZE);
    vector<float> h_B(SIZE);
    vector<float> h_C(SIZE);

    // Initialize matrices on the host
    for (int i=0; i<N; i++){
        for (int j=0; j<N; j++){
            h_A[i*N+j] = sin(i);
            h_B[i*N+j] = cos(j);
        }
    }

    // Allocate memory on the device
    dev_array<float> d_A(SIZE);
    dev_array<float> d_B(SIZE);
    dev_array<float> d_C(SIZE);

    d_A.set(&h_A[0], SIZE);
    d_B.set(&h_B[0], SIZE);

    matrixMultiplication(d_A.getData(), d_B.getData(), d_C.getData(), N);
    cudaDeviceSynchronize();

    d_C.get(&h_C[0], SIZE);
    c

In [19]:
# compilation test
bas = time.time()
!nvcc matrixmul.cu
son = time.time()

In [21]:
"2. test - compilation test result : " + str(son-bas)

'2. test - compilation test result : 2.9613592624664307'

In [22]:
# run the compiled file, test
bas = time.time()
!./a.out
son = time.time()

END

In [23]:
"2. test - time to run the compiled file : "+str(son-bas)

'2. test - time to run the compiled file : 8.686630487442017'

**1. Test Result : 1071.427838563919 (second)**

**2. Test Reult - Compile : 2.9613592624664307 (second)**

**2. Test Reult - run file : 8.686630487442017 (second)**